src/os/filestore/HashIndex.cc

```diff
@@ -552,15 +552,23 @@ int HashIndex::_created(const vector<string> &path,
     dout(1) << __func__ << " " << path << " has " << info.objs
             << " objects, starting split in pg " << coll() << "." << dendl;
     int r = initiate_split(path, info);
-    if (r < 0)
-      return r;
-    r = complete_split(path, info);
-    dout(1) << __func__ << " " << path << " split completed in pg " << coll() << "."
-            << dendl;
-    return r;
-  } else {
-    return 0;
+    if (r < 0) {
+      derr << __func__ << " error starting split " << path << " in pg "
+           << coll() << ": " << cpp_strerror(r) << dendl;
+      ceph_assert(!cct->_conf->filestore_fail_eio);
```

Comment:
what should we do when filestore_fail_eio is false? should we return the error as before?

This option only seems useful as a last resort to try to extract data from a failing disk anyway.

Reply:
No. We will not return the error -- just print an error message. I don't want the operation that triggered split/merge to be aborted in the middle due to this. The split/merge result does not affect the operation result, while if the op would be aborted in the middle due to this the returned error code may confuse the transaction (as it was in our case when  ENODATA was returned by the split; see the details in the tracker).

I have no a good idea if it would be useful to have a separate option to control the behavior. Sure, it would be less confusing (because we use it not only for EIO), but probably it is not worth doing. I am fine with any way, just wanted to mention it.

Reply:
I see what you mean about the lack of split/merge not affecting the op result. It seems reasonable as-is then. I certainly don't want to introduce more options for a tiny edge case.